api/src/service/cache_snapshot.ts

```diff
@@ -0,0 +1,413 @@
+  }
+
+  return { canPublish: true, eventData: publishEvent };
+}
+
+function getSourceInfo(
+  key: string,
+  eventType: string,
+): { searchKey: string; sourceFromSnapshot: Function; parseFromSnapshot: Function } {
+  let searchKey;
+  let sourceFromSnapshot: Function;
+  let parseFromSnapshot: Function;
+
+  switch (eventType) {
+    case PROJECT_SNAPSHOT_EVENT_KEY:
+      searchKey = SNAPSHOT_KEY;
+      sourceFromSnapshot = ProjectEventSourcing.sourceProjectFromSnapshot;
+      parseFromSnapshot = ProjectEventSourcing.parseProjectFromSnapshot;
+      break;
+    case SUBPROJECT_SNAPSHOT_EVENT_KEY:
+      searchKey = key + "_" + SNAPSHOT_KEY;
+      sourceFromSnapshot = SubprojectEventSourcing.sourceSubprojectFromSnapshot;
+      parseFromSnapshot = SubprojectEventSourcing.parseSubprojectFromSnapshot;
+      break;
+    case WORKFLOWITEM_SNAPSHOT_EVENT_KEY:
+      searchKey = key + "_" + SNAPSHOT_KEY;
+      sourceFromSnapshot = WorkflowitemEventSourcing.sourceWorkflowitemFromSnapshot;
+      parseFromSnapshot = WorkflowitemEventSourcing.parseWorkflowitemFromSnapshot;
+      break;
+    default:
+      // eslint-disable-next-line @typescript-eslint/explicit-function-return-type, @typescript-eslint/no-empty-function
+      return { searchKey: "", sourceFromSnapshot: () => {}, parseFromSnapshot: () => {} };
+  }
+  return {
+    searchKey: searchKey,
+    sourceFromSnapshot: sourceFromSnapshot,
+    parseFromSnapshot: parseFromSnapshot,
+  };
+}
+
+async function getItemsInfo(
+  conn: ConnToken,
+  ctx: Ctx,
+  streamName: string,
+  key: string,
+  searchKey: string,
+): Promise<Result.Type<{ items: Item[]; snapshotIndex: number }>> {
+  const rpcClient = conn.multichainClient.getRpcClient();
+  let items: Item[] = [];
+  try {
+    items = await rpcClient.invoke("liststreamkeyitems", streamName, key, false, MAX_ITEM_COUNT);
+    if (items.length == 0) {
+      return new VError("Data Not Found");
+    }
+  } catch (e) {
+    return new VError("Data Not Found");
+  }
+
+  // Traverse the list reverse to get latest snapshot
+  let snapshotIndex = -1;
+  for (let i = items.length - 1; i >= 0; i--) {
+    let item = items[i];
+    if (item.keys.includes(searchKey)) {
+      snapshotIndex = i;
+      break;
+    }
+  }
+  return { items: items, snapshotIndex: snapshotIndex };
+}
+
+export function parseBusinessEvents(
+  items: Item[],
+  streamName: string,
+): Array<Result.Type<BusinessEvent>> {
+  return items
+    .map((item) => {
+      const event = item.data.json;
+      if (!event) {
+        logger.warn(`ignoring event no item.data.json property found ${JSON.stringify(item)}`);
+        return;
+      }
+      if (event.intent) {
+        logger.warn(`ignoring event of intent ${event.intent}`);
+        return;
+      }
+      const parser = EVENT_PARSER_MAP[event.type];
+      if (parser === undefined) {
+        const eventType = event && event.type ? event.type : JSON.stringify(event);
```

Comment:
Condition 'event' is always true at this point because the false branch of the condition '!event' at line 343 has been taken.